helper_test.go

```diff
@@ -185,10 +209,72 @@ func echoTCPIdent(tt *tunneltest.TunnelTest, echo, ident string) (string, error)
 	}
 }
 
+func websocketDial(tt *tunneltest.TunnelTest, ident string) (*websocket.Conn, error) {
+	req := tt.Request(ident, nil)
+	if req == nil {
+		return nil, fmt.Errorf("no client found for ident %q", ident)
+	}
+
+	h := http.Header{"Host": {req.Host}}
+	wsurl := fmt.Sprintf("ws://%s", tt.ServerAddr())
+
+	conn, _, err := dialer.Dial(wsurl, h)
+	return conn, err
+}
+
 func sleep() {
 	time.Sleep(time.Duration(rand.Intn(2000)) * time.Millisecond)
 }
 
+func handlerEchoWS(t *testing.T) func(w http.ResponseWriter, r *http.Request) {
+	return handlerEchoSleepWS(t, false)
+}
+
+func handlerLatencyEchoWS(t *testing.T) func(w http.ResponseWriter, r *http.Request) {
+	return handlerEchoSleepWS(t, true)
+}
+
+func handlerEchoSleepWS(t *testing.T, doSleep bool) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		conn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
```

Comment:
in most of the cases i dont use both http error responses and test failures at the same time. returning the error as handler response and checking the return value in the calling function seems cleaner to me.